src/Tester/TestStreamProviders/Generator/GeneratorPooledCache.cs

```diff
@@ -0,0 +1,197 @@
+            private readonly Action<IDisposable> purgeAction;
+            private FixedSizeBuffer currentBuffer;
+
+            public CacheDataAdapter(IObjectPool<FixedSizeBuffer> bufferPool, Action<IDisposable> purgeAction)
+            {
+                if (bufferPool == null)
+                {
+                    throw new ArgumentNullException("bufferPool");
+                }
+                if (purgeAction == null)
+                {
+                    throw new ArgumentNullException("purgeAction");
+                }
+                this.bufferPool = bufferPool;
+                this.purgeAction = purgeAction;
+            }
+
+            public void QueueMessageToCachedMessage(ref CachedMessage cachedMessage, GeneratedBatchContainer queueMessage)
+            {
+                cachedMessage.StreamGuid = queueMessage.StreamGuid;
+                cachedMessage.StreamNamespace = queueMessage.StreamNamespace;
+                cachedMessage.SequenceNumber = queueMessage.RealToken.SequenceNumber;
+                cachedMessage.Payload = SerializeMessageIntoPooledSegment(queueMessage);
+            }
+
+            // Placed object message payload into a segment from a buffer pool.  When this get's too big, older blocks will be purged
+            private ArraySegment<byte> SerializeMessageIntoPooledSegment(GeneratedBatchContainer queueMessage)
+            {
+                // serialize payload
+                byte[] serializedPayload = SerializationManager.SerializeToByteArray(queueMessage.Payload);
+                int size = serializedPayload.Length;
+
+                // get segment from current block
+                ArraySegment<byte> segment;
+                if (currentBuffer == null || !currentBuffer.TryGetSegment(size, out segment))
+                {
+                    // no block or block full, get new block and try again
+                    currentBuffer = bufferPool.Allocate();
+                    currentBuffer.SetPurgeAction(purgeAction);
+                    // if this fails with clean block, then requested size is too big
+                    if (!currentBuffer.TryGetSegment(size, out segment))
+                    {
+                        string errmsg = String.Format(CultureInfo.InvariantCulture,
+                            "Message size is to big. MessageSize: {0}", size);
+                        throw new ArgumentOutOfRangeException("queueMessage", errmsg);
+                    }
+                }
+                Buffer.BlockCopy(serializedPayload, 0, segment.Array, segment.Offset, size);
+                return segment;
+            }
+
+            public IBatchContainer GetBatchContainer(ref CachedMessage cachedMessage)
+            {
+                //Deserialize payload
+                var stream = new BinaryTokenStreamReader(cachedMessage.Payload.ToArray());
+                object payloadObject = SerializationManager.Deserialize(stream);
```

Comment:
use the BinaryTokenStreamReader constructor overload that takes a list of ArraySegment (or even create an overload that takes just 1 segment), instead of using the Enumerable.ToArray() method